fix(ivy): run annotations handlers' `resolve()` in `ngcc`

gkalpak opened the pull request:
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~


## PR Type
- [x] Bugfix


## What is the current behavior?
The `resolve` phase (run after all handlers have compiled) was introduced in 7d954df, but `ngcc` was not updated to run the handlers' `resolve()` methods. As a result, certain operations (such as listing directives used in component templates) would not be performed by `ngcc`.


## What is the new behavior?
This PR fixes it by running the `resolve()` methods once compilation has been completed.


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


## Other information
Jirs issue: [FW-1149][1]

[1]: https://angular-team.atlassian.net/browse/FW-1149

petebacondarwin commented on packages/compiler-cli/src/ngtsc/imports/src/emitter.ts:
```diff
@@ -54,8 +54,8 @@ export interface ReferenceEmitStrategy {
  *
- * A `ReferenceEmitter` uses one or more `ReferenceEmitStrategy` implementations to produce a
- * an `Expression` which refers to a `Reference` in the context of a particular file.
+ * A `ReferenceEmitter` uses one or more `ReferenceEmitStrategy` implementations to produce an
```
😍

petebacondarwin reviewed (changes requested):
Looks like this breaks real-life - https://circleci.com/gh/angular/angular/220656 and https://circleci.com/gh/angular/angular/220659#tests/containers/2.

Also there is a minor linting issue.

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/analysis/decoration_analyzer_spec.ts:
```diff
@@ -92,8 +92,9 @@ describe('DecorationAnalyzer', () => {
-    beforeEach(() => {
-      const {options, host, ...bundle} = makeTestBundleProgram([TEST_PROGRAM]);
+    // Helpers
+    const setUpAndAnalyzeProgram = (...progArgs: Parameters<typeof makeTestBundleProgram>) => {
```
Fancy TS-foo 😛

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/analysis/decoration_analyzer_spec.ts:
```diff
@@ -104,7 +105,9 @@ describe('DecorationAnalyzer', () => {
-    });
+    };
+
+    beforeEach(() => setUpAndAnalyzeProgram(TEST_PROGRAM));
```
Perhaps this `beforeEach` and the following `it` clauses should be put in a describe block, otherwise we are running `setupAndAnalyzeProgram()` twice for the tests in `describe('internal components')

gkalpak replied:
Ha ha, yeah, I really wanted to do that, but couldn't come up with a good description 😇
I'll make something up 🤔

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/analysis/decoration_analyzer_spec.ts:
```diff
@@ -70,11 +70,16 @@ const INTERNAL_COMPONENT_PROGRAM = [
 ];
 
+type DecoratorHandlerWithResolve = DecoratorHandler<any, any>& {
+  resolve: NonNullable<DecoratorHandler<any, any>['resolve']>;
```
🥊 yet more TS-fu :-)

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -198,6 +200,16 @@ export class DecorationAnalyzer {
     return compilations;
   }
+
+  protected postCompileFile(compiledFile: CompiledFile): void {
```
`postCompileFile()` is a bit generic, given that this function does only one thing, which is `resolveDecoratedClasses()`?

gkalpak replied:
`resolve()` felt equally ambiguous, but at least using it would be consistent :grin:
I will rename it to `resolve()` (to also match [ngtsc][1]).

[1]: https://github.com/angular/angular/blob/1d88c2bb81b077efbf073318add678be2b035f4d/packages/compiler-cli/src/ngtsc/transform/src/compilation.ts#L245

petebacondarwin reviewed (approved):
A few minor comments that you may choose to implement - assuming that the caretaker doesn't auto-merge it before you read my review :-)

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/analysis/decoration_analyzer_spec.ts:
```diff
@@ -78,22 +97,34 @@ function createTestHandler() {
+  handler.compile.and.callFake((decl: ts.Declaration, analysis: any) => {
+    expect(handler.resolve).toHaveBeenCalledWith(decl, analysis);
+    return analysis.resolved && `@${analysis.decoratorName} (compiled)`;
+  });
```
Personal opinion: I find this method of detecting the order of things awkward to follow. IMO it is cleaner to log the calls to an array (or similar) and then have an expectation that checks the array has the items in the correct order.

gkalpak replied:
Agreed 👍
I'll change it.

mary-poppins commented:
You can preview 407c9de at https://pr28963-407c9de.ngbuilds.io/.
You can preview 709cd89 at https://pr28963-709cd89.ngbuilds.io/.
You can preview 76e4716 at https://pr28963-76e4716.ngbuilds.io/.
You can preview 572769c at https://pr28963-572769c.ngbuilds.io/.
You can preview eda33d7 at https://pr28963-eda33d7.ngbuilds.io/.
You can preview f34fae5 at https://pr28963-f34fae5.ngbuilds.io/.
You can preview f29b4e8 at https://pr28963-f29b4e8.ngbuilds.io/.
You can preview ebcdafb at https://pr28963-ebcdafb.ngbuilds.io/.
You can preview 9720920 at https://pr28963-9720920.ngbuilds.io/.
You can preview bc6680e at https://pr28963-bc6680e.ngbuilds.io/.
You can preview 737636f at https://pr28963-737636f.ngbuilds.io/.
You can preview d41e2a8 at https://pr28963-d41e2a8.ngbuilds.io/.
You can preview 1167166 at https://pr28963-1167166.ngbuilds.io/.

mary-poppins commented:
You can preview a38c314 at https://pr28963-a38c314.ngbuilds.io/.

kara commented:
[presubmit](http://test/OCL:238909181:BASE:238909323:1552871483967:c3916e)

gkalpak committed:
refactor(ivy): remove unnecessary escaping in RegExp

gkalpak committed:
refactor(ivy): improve error message in `ngtsc`'s `findExportedNameOfNode()`

gkalpak committed:
refactor(ivy): avoid code duplication in `ngcc` tests

gkalpak committed:
test(ivy): expand `ngcc` `DecorationAnalyzer` tests to cover more cases

gkalpak committed:
fixup! fix(ivy): run annotations handlers' `resolve()` in `ngcc`

gkalpak committed:
fix(ivy): run annotations handlers' `resolve()` in `ngcc`

The `resolve` phase (run after all handlers have analyzed) was
introduced in 7d954dffd, but `ngcc` was not updated to run the handlers'
`resolve()` methods. As a result, certain operations (such as listing
directives used in component templates) would not be performed by
`ngcc`.

This commit fixes it by running the `resolve()` methods once analysis
has been completed.

gkalpak committed:
fixup! fix(ivy): run annotations handlers' `resolve()` in `ngcc`

mary-poppins commented:
You can preview 778de01 at https://pr28963-778de01.ngbuilds.io/.
